doc/design/draft-metadata-migration.md

```diff
@@ -0,0 +1,13 @@
+Premise: Draft metadata should always be in the latest schema version.
+
+There are a few places in the API where draft metadata are altered (POST, PUT and served (GET). Note that the UI and 
+API use a shared schema version. All actions should thus assume that this is the only version available to the UI and
+server. Any other versions need to be migrated to this version for any display or editing.
+
+1. The API should call dandischema `migrate` on all metadata exchanges of draft metadata.
```

Comment:
Does this mean that if a client tries to PUT/POST metadata with `schemaVersion=0.2.0` for example, then the server would automatically `migrate` it to whatever the current `schemaVersion` is before saving, and throw an error if migration fails?
Or does it mean that any draft metadata stored in the DB with an out of date `schemaVersion` will be automatically `migrate`d on every `GET` so that it appears to be up to date?

Reply:
I vote we only migrate on incoming data:
```suggestion
1. The API should call dandischema `migrate` on all incoming metadata draft metadata, and throw an error if migration fails. This ensures that the metadata in the DB will always have an up to date schemaVersion.
```
If we also migrate all drafts immediately after incrementing the schemaVersion, the DB is guaranteed to always be in a good state.

Reply:
this would not provide such a guarantee since if dandi-api starts to use newer dandischema (library/schema) then all dandiset metadata will be outdated in DB... that is why I would have preferred to

```suggestion
1. The API should provide `PUT /dandisets/{dandiset_pk}/migrate/` endpoint, which would return migrated version, and description of changes which happened, warnings.
1a. The API would refuse PUT/POST of incompatible metadata.
```

`1a` is similar to how `dandi-cli` refuses to upload an asset with non-matching metadata version ATM. More comprehensive support of non-exact versions upload is proposed in https://github.com/dandi/dandi-api/issues/437

Reply:
I agree with Satra's philosophy:
> these are the kinds of things i would prefer to take care of behind the scenes. users should not have to think about this.

As long as we don't allow publishing with an out of date schema version, I don't see a problem with having schemaVersion being outdated briefly while the migration script is being run.

1. Someone upgrades the `schemaVersion` of the server. This triggers a redeploy of the server.
  Anyone who tries to publish a dandiset after the deployment will find the button and the endpoint disabled because the schema version is out of date.  
2. Whoever upgraded the `schemaVersion` runs the `migrate_version_metadata` script, which upgrades the schema version of all the draft dandisets.

Reply:
I would agree with auto- behind the scenes things, which are **guaranteed** not to fail/transform or loose (meta)data user entered etc. I do not think we have such a guarantee for metadata migrations. And for any migration support now we will need to pretty much grow the chain of transformations, which would become more and more "intrusive".  If all those happen behind the scene, and I as a user come and do not see metadata I have entered where I have entered it -- I might get confused etc.
Having said that DLP for read-only display could auto-migrate and add a notice that "Metadata was auto-migrated from version X for display". Needed anyways for all published dandisets since we would not be automigrating them, right? 
Then owner(s) upon publish would acknowledge the correctness of migration via edit/save and having been displayed relevant information.

> 2\. Whoever upgraded the `schemaVersion` runs the `migrate_version_metadata` script, which upgrades the schema version of all the draft dandisets.

and introspects correct migration for all 100s of dandisets, or just does not bother and thus possibly loosing data (I think it is not unprecedented to have bugs in the code) without a way to "redo it" (migrated metadata replaces original right?)?

Reply:
situation BTW is not unlike we are asking pynwb folks to address -- to be able to see data in the original (as in the file) version of schema, since all the automigration during opening without being able to load as it was originally saved did lead to surprising results and errors IIRC. Here also if we auto-migrate draft metadata during "long draft period", any user script which uses metadata for anything, would need to be constantly and immediately updated shortly after each migration, thus annoying the user/developer.

Reply:
i would like to dissociate pynwb use case from this one. i do not think they are the same (dandischema would be). 

- from the API and  UI perspective both can use a single schema. 
- from a draft perspective, any usage of draft metadata by a user/script is not guaranteed. so if they are coding they should not rely on draft metadata being stable whether long or short
- in terms of migration, we would always be doing a one version (patch, minor, or release), it would always jump from the last schema, thus there should not be any process that should fail unless we decide something was ambiguous and cannot be transitioned. this would then be reflected in the schema validation message. "could not migrate, please contact help. etc.,."
 - for any already published dandiset, we can check what a new version of the schema will do for migration, and our migration code should account for that in dandischema.

Reply:
To get the ball rolling again, how about this @yarikoptic ?
```suggestion
1. The API should provide `GET /dandisets/{dandiset_pk}/versions/{version}/migrated/` endpoint, which would return the migrated metadata of the version, and a description of changes and warnings.
   For up to date metadata, this is obviously a no-op.
   The meditor would use this endpoint to fetch the draft metadata whenever it is opened, which ensures that editing a draft with out of date schema version will implicitly migrate it.
   This endpoint can also be used to get the metadata of published versions as if they were migrated.
2. The API will not accept metadata with out of date `schemaVersion`.
   Draft metadata with out of date `schemaVersion` is considered invalid by the API (even if it would pass validation) and is not publishable.
```

Instead of having `PUT /dandisets/{dandiset_pk}/migrate/` do the migration and save it, `GET /dandisets/{dandiset_pk}/versions/{version}/migrated/` does the migration, which is then displayed in the meditor for the owner's awareness, and only saved if it's cool with them.

Now when we bump the schema version, we do not have to migrate anything behind the scenes. All drafts are immediately unpublishable. Migration would be done by the owners by opening the meditor and saving it. In the event of migration errors, the meditor would show those to the owner at that time.

Reply:
Paging Dr. @yarikoptic ^^^

Reply:
Re 1. explicit `/migrated/` -- sounds good .

Re 2. "not accept metadata with out of date..." -- for *now* I think it is ok, but I think we might still end up needing to https://github.com/dandi/dandi-api/issues/437 to not demand some "metadata upgrade" whenever newer schema .patch version just fixes  something trivial and non-consequential (e.g. documentation etc).

Reply:
When we implement #437 we can change "not out of date" to mean "within a minor version", or whatever makes sense at that time. I don't see any problem augmenting this design at a later date, so I'm going to apply this suggestion.